frontend/packages/sdk/src/contracts/PinkContract.ts

```diff
@@ -0,0 +1,315 @@
+import type { Registry } from '@polkadot/types/types';
+import type { DecorateMethod, ApiTypes } from '@polkadot/api/types';
+import type { KeyringPair } from '@polkadot/keyring/types';
+
+import type { OnChainRegistry } from '../OnChainRegistry';
+import type { AbiLike } from '../types';
+import type { CertificateData } from '../certificate';
+
+import { Abi } from '@polkadot/api-contract/Abi';
+import { toPromiseMethod } from '@polkadot/api';
+import { ContractSubmittableResult } from '@polkadot/api-contract/base/Contract';
+import { applyOnEvent } from '@polkadot/api-contract/util';
+import { withMeta, convertWeight } from '@polkadot/api-contract/base/util'
+import { BN, BN_ZERO, hexAddPrefix, u8aToHex, hexToU8a } from '@polkadot/util';
+import { sr25519Agree, sr25519KeypairFromSeed, sr25519Sign } from "@polkadot/wasm-crypto";
+import { from } from 'rxjs';
+
+import { pruntime_rpc as pruntimeRpc } from "../proto";
+import { decrypt, encrypt } from "../lib/aes-256-gcm";
+import { randomHex } from "../lib/hex";
+
+
+export interface ContractInkQuery<ApiType extends ApiTypes> extends MessageMeta {
+  (origin: KeyringPair, ...params: unknown[]): ContractCallResult<ApiType, ContractCallOutcome>;
+}
+
+export interface MapMessageInkQuery<ApiType extends ApiTypes> {
+  [message: string]: ContractInkQuery<ApiType>;
+}
+
+class PinkContractSubmittableResult extends ContractSubmittableResult {
+
+  readonly #registry: OnChainRegistry
+
+  #isFinalized: boolean = false
+
+  constructor(registry: OnChainRegistry, result: ISubmittableResult, contractEvents?: DecodedEvent[]) {
+    super(result, contractEvents)
+    this.#registry = registry
+  }
+
+  async waitFinalized(timeout: number = 120_000) {
+    if (this.#isFinalized) {
+      return
+    }
+
+    if (this.isInBlock || this.isFinalized) {
+      const codeHash = this.status.asInBlock.toString()
+      const block = await this.#registry.api.rpc.chain.getBlock(codeHash)
+      const chainHeight = block.block.header.number.toNumber()
+
+      const t0 = new Date().getTime();
+      while (true) {
+        const result = await this.#registry.phactory.getInfo({})
+        if (result.blocknum > chainHeight) {
+          this.#isFinalized = true
+          return
+        }
+
+        const t1 = new Date().getTime();
+        if (t1 - t0 > timeout) {
+          throw new Error('Timeout')
+        }
+        await new Promise(resolve => setTimeout(resolve, 1000));
+      }
+    }
+    throw new Error('Contract transaction submit failed.')
+  }
+}
+
+
+function createQuery(meta: AbiMessage, fn: (origin: string | AccountId | Uint8Array, options: CertificateData, params: unknown[]) => ContractCallResult<'promise', ContractCallOutcome>): ContractInkQuery<'promise'> {
+  return withMeta(meta, (origin: string | AccountId | Uint8Array, options: CertificateData, ...params: unknown[]): ContractCallResult<'promise', ContractCallOutcome> =>
```

Comment:
Here, `(origin, option, ...)` becomes `(origin, cert, ...)`. I'm wondering why.

In the previous version, `origin` means the sender. Vanilla ink doesn't have privacy. So everyone can pass the strong of the account id as the origin. In Phat Contract there's privacy and permission. So we pass a signed certificate instead.

In this version, do we need to pass both the origin and the certificate? Where's the options (to specify gas fee, etc)?

Reply:
The original Polkadot contract query API includes options such as `gasLimit` and `storageDepositLimit`, which are irrelevant for off-chain queries. Instead, users must provide a certificate. To reflect this change, we propose replacing the `options` parameter with `cert`. Additionally, we suggest making the handling of certificate caching user-controlled rather than internal. Initially, we attempted to handle certificate signing automatically but found that it caused pop-ups in the browser asking users to sign.

```
contractPromise.query.foo(address, options, ....args)
// Become
contractPromise.query.foo(address, cert, ....args)
```

Reply:
In the context of Phat Contract, the original `address` is irrelevant. The cert represents the identity. So I think it makes more sense to replace address by the cert. On the other hand, the options is not used. So probably we can just leave it as an empty struct. Sample would look like:

```typescript
interface PhatContractOption {}
contractPromise.query.foo(cert, {}, ....args)
```

Reply:
> In the context of Phat Contract, the original `address` is irrelevant. The cert represents the identity. So I think it makes more sense to replace address by the cert. On the other hand, the options is not used. So probably we can just leave it as an empty struct. Sample would look like:
> 
> ```ts
> interface PhatContractOption {}
> contractPromise.query.foo(cert, {}, ....args)
> ```

This is what current API looks like:

```ts
const aliceCert = await signCertificate({ pair: alice, api })
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(alice, aliceCert, 'Badge#1')
const options = {
   gasLimit: gasRequired.refTime,
   storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

If we skip the address:

```ts
const aliceCert = await signCertificate({ pair: alice, api })
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(aliceCert, {}, 'Badge#1')
const options = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

The original API:

```ts
const { gasRequired, storageDeposit, result } = await contract.query.newBadge(alice, { storageDepositLimit: null, gasLimit: -1 }, 'Badge#1');
const options = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contract.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

Or we can re declare the options as:
```ts
interface PhatContractQueryOption {
  cert: any
}

const queryOptions = await Phala.getQueryOptions(alice)
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(alice, options, 'Badge#1')
const txOptions = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(txOptions, 'Badge#1').signAndSend(alice)
```

Then we can get more similar API.

Reply:
Let me give another example show the benefit of keeping similar API:

```ts
const api = ApiPromise.create({ ... })
const DIProvider = new PolkadotCompatibleContractServiceProvider(api)
//
// We don't care about it's original Polkadot JS ContractPromise instance or PinkContractPromise instance
//
const contractPromise = await DIProvider.createContractPromise(contractId)
const queryOptions = awiat DIProvider.createContractQueryOptions(alice)
await contractPromise.query.newBadge(alice, queryOptions, 'Badge #1')
```